152/change trigger for setting implicit location

shen-sat opened the pull request:
Hi @JacobEvelyn, hope you are having a good Xmas! 🎄

I've made the changes [we discussed](https://github.com/JacobEvelyn/friends/issues/152):
- changed trigger for setting a default location from `moved to _LOCATION_` to `to _LOCATION_`
- I've added functionality (plus tests) to print `Default location set to:  _LOCATION_` to stdout when a `to _LOCATION_` activity is added by the user
- I've updated the readme for setting default locations

While making this PR I came across another instance of us using `to _LOCATION_` in a test where we don't want to set a default location (this is expected as `to _LOCATION_` wasn't used before as a trigger for anything). I removed/amended it, but it did raise the question again of why we are using `to _LOCATION_` as the trigger when its such a commonly used phrase, and why we are not using something more explicit (like `moved to` or `set default location to`). I can easily imagine users writing `I went to Paris for the day` or even `I went to Paris` without wanting to set the default location to Paris.

Just a minor re-push-back from me :) Let me know what you think, happy to discuss changes further 👍 

PS is there a way to run one test a time with minitest? I had a look online, but couldn't find a method that didn't involve installing another gem 🤔  

---

Hi there! Thanks so much for submitting a pull request!

Let's just make sure together that all of these boxes are checked before we
merge this change:

- [x] The code in these changes works correctly.
- [x] Code has tests as appropriate.
- [x] Code has been reviewed by @JacobEvelyn.
- [x] All tests pass on Travis CI.
- [x] Code coverage remains high.
- [x] Rubocop reports no issues on Travis CI.
- [x] The `README.md` file is updated as appropriate.

Don't worry—this list will get checked off in no time!

JacobEvelyn commented:
Thanks so much for making this PR, @shen-sat ! I totally hear your concerns about an overly sensitive trigger for this. If you're okay with it, let me think on this a bit more. In the meantime, I'll review your PR as-is and leave a few comments!

JacobEvelyn commented on README.md:
```diff
@@ -141,7 +142,7 @@ FRIENDS_PAGER="less -RF"
 - **Friends**: The people you do _activities_ with. Each friend has a name and,
   optionally, one or several nicknames. (Examples: `John`, `Grace Hopper`)
 - **Locations**: The places in which _activities_ happen. (Examples: `Paris`,
-  `Marie's Diner`)
```
I think this is a good change regardless of the broader discussion. This example, however, intended to show that locations can be multiple words and contain punctuation. Could we switch to a relatively-known one that still fits that description? Maybe `Martha's Vineyard` or something similar? (I'm having trouble finding other examples, eek.)

shen-sat replied:
Coolio - in that case, keeping Marie's Diner as an example here is fine, no? If we are not worried about someone typing in `Went to _Marie's Diner_` and accidentally setting `Marie's Diner` as a default location (for the moment), then I'm happy to keep this example as is 👍

JacobEvelyn replied:
I'm not totally sure I understand what you're saying. I think I'd like to change this to a geographic location that still uses both spaces and punctuation (`Martha's Vineyard`, for example) to show users that locations don't need to be a single word. Does that make sense?

I think I may be misunderstanding you—please correct me if so!

shen-sat replied:
Ah gotcha, this makes sense - I didn't know Martha's Vineyard was a place (I thought it might be a wine bar or something 😂) so happy to replace `Marie's Diner` with `Martha's Vineyard`. It's funny, I just tried Googling for a city/region with an apostrophe, and `Martha's Vineyard` kept coming up!

JacobEvelyn commented on README.md:
````diff
@@ -420,6 +420,31 @@ $ friends add nickname "Grace Hopper" "Amazing Grace"
 Nickname added: "Grace Hopper (a.k.a. The Admiral a.k.a. Amazing Grace)"
 ```
 
+#### Add a default location
````
I have a slight preference for "Setting a default location" but am open to being convinced otherwise! Thoughts?

shen-sat replied:
Your suggestion is waaay more descriptive, I'm with you :) 👍

JacobEvelyn commented on friends.md:
```diff
@@ -1,5 +1,5 @@
 ### Activities:
-- 2018-11-01: **Grace Hopper** and I went to _Marie's Diner_. George had to cancel at the last minute. @food
+- 2018-11-01: **Grace Hopper** and I went to Marie's Diner. George had to cancel at the last minute. @food
```
Maybe: `Marie's Diner on _Martha's Vineyard_`?

shen-sat replied:
Ah I see what you're doing here - I take it `Martha's Vineyard` is somewhere I could move to (ie set as a default location)? In which case, we could just change the example to `...and I went to _Martha's Vineyard_`, right? There's no need for `Marie's Diner` because it is not somehwere I could move to and we are already fulfilling the example of a location that is made up of multiple words with `_Martha's Vineyard_`

JacobEvelyn replied:
Yes, exactly! Sorry I couldn't think of a better location that fits these criteria. I almost did `St. Kitts` but technically the name is `Saint Kitts` and I wanted to use punctuation too. 🙁

shen-sat replied:
Shall I change all instances of `_Marie's Diner`_ to `_Martha's Vineyard_`? Or shall I change them all to `Marie's Diner on _Martha's Vineyard_`?

JacobEvelyn replied:
I think either way is fine! Totally your call.

JacobEvelyn commented on lib/friends/introvert.rb:
```diff
@@ -110,6 +110,9 @@ def add_activity(serialization:)
         @activities.unshift(activity)
 
         @output << "Activity added: \"#{activity}\""
+        if activity_sets_default_location?(activity)
```
I'm not sure I understand the rationale for the `previous_activities.none?(&:default_location)` check in this method. Why can't this simply be:

```ruby
if activity.default_location
  @output << "Default location set to: \"#{activity.default_location}\""
end
```

?

shen-sat replied:
True, I originally thought the same thing. But  after some thinking I came across the scenario covered by [this test](https://github.com/shen-sat/friends/blob/c108c5de5fb9c7908bc73537d59a47c56d56ea22/test/commands/add/activity_spec.rb#L81).

The first time I set a default location (eg with the line `Went to _Paris_ for a holiday`), then I would want to have the line `Default location set to Paris` printed to the console. But if I add a subsequent activity like `Moved to _Paris_` (which also contains the trigger for setting a default location to `Paris`), then printing `Default location set to Paris` is unneccessary as we've already notified the user with the same message - and this is what my implementation prevents. What do you think? Happy to go with your recommendation on this!

shen-sat replied:
Hmm, reading my own comment has just raised an issue I didn't think of before. If someone adds activities in this order:

1. `Moved to Paris`
2. `Moved to Berlin`
3. `Went to Paris`

Then when I add the third activity, with my implementation, `friends` **won't** print `Default location set to Paris` :(

In which case your implemention is correct, as long as we are both happy with `Default location set to _LOCATION_` being printed to console twice if someone adds activities in the follwing order:

1. `Moved to Paris`
2. `Went to Paris`

What do you think?

JacobEvelyn replied:
Oh! I see what you're saying—nice attention to detail.

Yes, the `to Paris`, `to Berlin`, `to Paris` example was what I was thinking of when I originally commented. Would you mind adding that as a test to make sure we _do_ print the message in that case?

I hadn't thought of the case of `to Paris`, `to Paris`, but I think you're right that it's weird to print the normal message. I actually worry about _not_ printing something though, because if I forget that the default location was set to that location already, I'd expect output and get nothing and would think something was wrong. I think the best solution might be something like this:

```ruby
if activity.default_location
  # This condition isn't quite right because it doesn't handle the case where activity is
  # inserted in the middle of our history... see below
  if @activities.drop(1).find(&:default_location).default_location == activity.default_location
    @output << "Default location already set to: \"#{activity.default_location}\""
  else
    @output << "Default location set to: \"#{activity.default_location}\""
  end
end
```

But now this is making me realize we should probably add tests for when the added activity is older and there's a newer default location. In that case I think it might be most helpful to print something like:

> Default location from **2016-01-01** to **2016-03-12** set to: "Paris"

Which is making me think we should maybe have a few different cases, something like:

```ruby
# this is all pseudocode
# note that some of these lookups may be easier to do in the `clean` method; I'm not sure.
if activity.default_location
  str = "Default location"

  # If we find one, keep searching back to make sure there isn't an earlier one.
  earlier_activity = activity before `activity` that has same default location || activity

  if activity is NOT latest activity
    str += " from #{Paint[earlier_activity.date, :bold]}"

    if later_activity changes the default location # If later_activity sets it to the same, keep searching for another later_activity that changes it
      str += " to #{Paint[later_activity, :bold]}"
    end
  end

  str += " already" if earlier_activity != activity

  @output << "#{str} set to: \"#{activity.default_location}\""
end
```

Thoughts?

shen-sat replied:
This is really interesting - I hadn't thought about the fact that a user could put in an _older_ activity! 😲

So from our discussion, there are 4 scenarios to cover:

1. User adds activity `Moved to _Paris_` with an empty activities log
2. User adds activity `Moved to _Paris_`, with `Went to _Paris_` as the last activity in the activities log
3. User adds activity `Moved to _Paris_`, with `Went to _Berlin_` as the last activity and `Went to _Paris_` as the activity before that in the activities log
4. User adds activity `Moved to _Paris_`, with a date that precedes the date of `Went to _Paris_` in the activities log

I think the pseudo code you wrote (thanks for writing it, it helped me understand the scenarios 😄 ) covers everything but scenario 3, but I can figure that out.

If I test those scenarios, do you think we'll have everything covered?

JacobEvelyn replied:
Thanks for writing these out! Good call on case 3 not being covered. Maybe a clearer version would be:

```
# this is still psuedocode!

# We could think about moving this stable_sort to the read stage
# so we don't also have to do it in `clean`. Or we could skip it since
# @activities will only be unsorted if the user has been editing the
# file manually.
@activities = stable_sort(@activities)

if activity.default_location
  str = "Default location"

  # We use <= instead of < here because a newly-added activity is guaranteed to be
  # the latest of all the activities on that date.
  earlier_activities, later_activities = @activities.partition { |a| a.date <= activity.date }

  earlier_activity = activity
  earlier_activities.reverse_each do |a|
    next unless a.default_location

    if a.default_location == activity.default_location
      earlier_activity = a
    else
      break
    end
  end

  # We use < instead of <= here because a newly-added activity is guaranteed to be
  # the latest of all the activities on that date.
  if !later_activities.empty? && activity.date < later_activities.last.date
    str += " from #{Paint[earlier_activity.date, :bold]}"

    later_activity = later_activities.find { |a| a.default_location && a.default_location != activity.default_location }

    # I added the " to present" case just now as I think it makes some of the output clearer.
    str += " to #{Paint[later_activity&.date || 'present', :bold]}"
  end

  str += " already" if earlier_activity != activity

  @output << "#{str} set to: \"#{activity.default_location}\""
end

@activities.unshift(activity) # Add this after we've done our lookups above.
```

So in terms of scenarios to test, I find it easiest to think about in terms of the output branches. We want tests that can produce output for all of the following formats:

> Default location set to: "Paris"
> Default location already set to: "Paris"
> Default location from **2016-01-01** to **present** set to: "Paris"
> Default location from **2016-01-01** to **present** already set to: "Paris"
> Default location from **2016-01-01** to **2017-01-01** set to: "Paris"
> Default location from **2016-01-01** to **2017-01-01** already set to: "Paris"

Thoughts?

shen-sat replied:
I haven't looked through the pseudo-code properly yet (I will do soon!), but I agree with thinking about producing all the outputs you listed. The only thing to add is that the first item you listed (`Default location set to: "Paris"`) will cover scenarios 1 and 3 from my list - I can make this explicit in the tests 👍

JacobEvelyn commented on README.md:
```diff
@@ -420,6 +420,31 @@ $ friends add nickname "Grace Hopper" "Amazing Grace"
 
+#### Add a default location
+
+You can set a default location in`friends`. Thereafter, whenever an activity is added and does not mention a location, the activity will be automatically 'tagged' with the default location.
```
I think we might be able to tighten this language up a bit. What are your thoughts on something like:

When an activity includes the phrase `to _LOCATION_` (e.g., `Took a plane to _Paris_`), all future activities that have no explicit location will be associated with that location:

```bash
$ friends add activity Took a plane to Paris
Activity added: "2020-01-04: Took a plane to Paris"
Default location set: "Paris"
$ friends add activity Ate lunch at a charming café
Activity added: "2020-01-04: Ate lunch at a charming café"
$ friends add activity Left the city to go to _Chamonix_
Activity added: "2020-01-04: Left the city to go to Chamonix"
```

```bash
$ friends list activities --in Paris
2019-01-04: Ate lunch at a charming café
2019-01-04: Took a plane to Paris
```

shen-sat replied:
I like your description - and this feature deserves explicit examples. I'll put your change in 👍

JacobEvelyn commented on test/commands/add/activity_spec.rb:
```diff
@@ -51,5 +51,38 @@
+    describe "when default location has not been set before" do
+      let(:activity) { "Had dinner in _Paris_" }
+
+      it 'prints "Default location added" output message' do
```
`added` --> `set to`

shen-sat commented:
> Thanks so much for making this PR, @shen-sat ! I totally hear your concerns about an overly sensitive trigger for this. If you're okay with it, let me think on this a bit more. In the meantime, I'll review your PR as-is and leave a few comments!

No worries @JacobEvelyn , it's always fun working on `friends` :)

I'm not too concerned about the trigger situation, I just wanted/thought it best to flag it to you as something that has organically come to my attention via contradictory examples the codebase itself provides - totally cool to roll with what you decide on this as we've discussed it quite a bit already 👍 

I do need a bit of clarification on why `Martha's Vineyard` is better than `Marie's Diner` though - I understand it is also made up of multiple words, but is it also a location someone can move to (and therefore set as a default location)? I've alo put in some other questions in my comments - once these are resolved I'll implement the required changes 😄 

PS did you see my previous question re: executing specific tests in `friends`? If you are able to do this, what command do you use?

PPS I just realised I didn't answer your question around my being a contributor to `friends` - _I think_ this might be to do with my choice to make my email private. Let me try making it public and seeing what happens 🤞

JacobEvelyn commented:
Hey @shen-sat ! Thanks as always for the thoughtful discussion and careful coding! Looking forward to merging this soon so you can take a stab at some more interesting features! (Or take a break from me if you need one. 😅)

> I'm not too concerned about the trigger situation, I just wanted/thought it best to flag it to you as something that has organically come to my attention via contradictory examples in the codebase itself - totally cool to roll with what you decide on this as we've discussed it quite a bit already 😄

I do hear your concern though! I think once we make this change we can see how it feels to use, and whether we get any feedback from users. As an example of something like this in the past, tagging in `friends` was originally with the `#` character as in Twitter. It soon became clear that this was the wrong decision (https://github.com/JacobEvelyn/friends/issues/122#issuecomment-219570641) and we fixed it. I'm more than happy to revise previous decisions when new evidence comes to light! (Plus, `friends` isn't yet at a `1.0` release, so I feel we have a bit more leeway in this department.)

> I do need a bit of clarification on why Martha's Vineyard is better than Marie's Diner though (see my comments, above). I understand Martha's Vineyard is also made up of multiple words, but is it also a location someone can move to (and therefore set as a default location)?

Ah, yes, sorry for the confusion! [Martha's Vineyard](https://en.wikipedia.org/wiki/Martha's_Vineyard) is an island in the northeastern United States—known there as a summer getaway for the wealthy. I couldn't think of any other examples of places that have both punctuation and spaces, but perhaps it would be best to use multiple locations to illustrate this instead?

> I've also put in some other questions in my comments - once these are resolved I'll implement the required changes 👍

Will take a look, thanks!

> PS did you see my previous question re: executing specific tests in friends? If you are able to do this, what command do you use?

Whoops, sorry for forgetting to respond to this! This is indeed an annoyance I have (for my day job I use `RSpec`, which is much more full-featured and makes this sort of thing easier), and I haven't found a good way to do it without adding a gem. (There's supposedly a way to specify the entire test name on the command line but it's a lot of trouble and I'm not sure I've ever gotten it to work correctly.) I added the steps I usually take to step 6 here: https://github.com/JacobEvelyn/friends/blob/master/.github/CONTRIBUTING.md

> PPS I just realised I didn't answer your question around my being a contributor to friends - I think this might be to do with my choice to make my email private. Let me try making it public and seeing what happens 🤞

I doubt your email needs to be public on your profile, but I found [this help article](https://help.github.com/en/github/setting-up-and-managing-your-github-profile/why-are-my-contributions-not-showing-up-on-my-profile) and it says the email you use in your commit (which I won't type here but you can see with `git log`) needs to be tied to your account. Could your account be using a different email?

JacobEvelyn commented:
Hey @shen-sat! Anything I can do to help get this over the line?

shen-sat commented:
Hi @JacobEvelyn soooo sorry for the late reply! I've been busy with other projects, and I haven't been keeping on top of my emails/notifications. Totally my bad :(

I'm heading to bed now, but I'll pick this up and (re)read through your comments tomorrow if that's cool? I'll shout if I have any questions

JacobEvelyn commented:
@shen-sat let me know if my answers make sense! Happy to discuss anything, as always.

shen-sat commented:
All makes sense, thanks @JacobEvelyn, I'll begin start the work 👌 

Whoa it's been downloaded over 39 thousand times, that's awesome! Yeah, it's interesting with gems - there isn't anything to measure/monitor (apart from downloads) unlike other apps like websites or mobile apps. If I have any suggestions for tracking use/behaviour, I'll give you a shout 👍

shen-sat commented:
Planning to have something for you to review this weekend 🤞

JacobEvelyn commented:
I noticed you pushed a few more commits @shen-sat! Just wanted to say let me know when you'd like me to review, and please let me know if there's anything I can help with!

shen-sat commented:
Hi @JacobEvelyn! I've been working on this over the weekend - it's a little more complicated than I thought it was going to be, but your pseudo code has really helped me structure the implementation.

I've got a lot of refactoring still to do: the code needs DRYing, and rubocop is complaining about some lines that are too long. I'd therefore like to clean up the code before you officially review it if that's ok?

However, one thing that would help me is if you reviewed specifically the tests now for two things:

1. Have I covered all the scenearios we discussed?
2. I've written private methods for the tests for the first time (because I don't think I can use shared examples with `minitest`) - do my private methods make sense?

If you can give me input for the above to questions, that would super awesome. I can then incorporate your comments while I refactor the rest of the code and then give you the official thumbs up to review 👍  And thanks for being so patient! :)

shen-sat commented:
Actually @JacobEvelyn - scratch that! I've just looked at my code after a night's sleep and had a eureka moment 💡 I've thought of a much much better way to re-arrange my tests so that they are a) comprehensive and b) way easier for you (and future devs) to read.

Would you mind holding off reviewing for the moment (ie ignoring my last comment)? I'm busy for the next couple of days, so I'd like to timebox this refactor until the end of this week. After that point, whether I've finished or not, I will have looked at the code for too long and thus it will be time for your review :)

Let me know if that's cool with you!

Would you mind

JacobEvelyn commented:
That works for me! ❤️

shen-sat commented:
Hey @JacobEvelyn! I'm finally done with this PR (for now).

I’m not _super_ satisfied with my implementation, but as I mentioned in my last comment I think it’s time for a fresh pair of eyes to look at it. Tests are passing, but it could do with some DRYing and rubocop is complaining about lines that are too long. I’ve also used very long (but descriptive!) names for variables and methods - I’ll need to change these too.

Deciding what scenarios to test was the most challenging part of this PR, but it was also really fun to try and figure out. It was actually quite hard to decide when to _stop_ writing tests, because there are potentially many different scenarios a user can find themselves in, especially when adding a default location in the past. I’ve tried to structure the tests like so:
- With the activity added being the latest activity
- With the activity added being an activity in the past:
    - With no proceeding default locations
        - Different preceding activity scenarios
    - With no preceding default locations
        - Different proceeding activity scenarios
    - With the preceding default location the same as the added activity
        - With same proceeding default location
        - With different proceding default location
    - With the preceding default location different to the added activity
        - With same proceeding default location
        - With different proceding default location

I look forward to your review - as alway, very happy to discuss/explain my implementation :)

shen-sat commented on lib/friends/introvert.rb:
```diff
@@ -777,5 +779,52 @@ def thing_with_name_in(type, text)
+    end
+
+    def earliest_consecutive_default_location_activity_date(activity, existing_activities)
+      # binding.pry
```
I need to delete this 😅

JacobEvelyn commented on lib/friends/introvert.rb:
```diff
@@ -777,5 +779,52 @@ def thing_with_name_in(type, text)
+
+    def earliest_consecutive_default_location_activity_date(activity, existing_activities)
+      # binding.pry
+      earlier_default_location_activities = existing_activities.select { |a| a.default_location && (a.date < activity.date) }
```
I don't think you need the parentheses here, right? Also, using a `partition` as in my version helps to avoid having to do these date comparisons in multiple places. Thoughts?

shen-sat replied:
Parentheses can definitely be removed, and yes using the `partition` method is much more efficient 👍

JacobEvelyn commented on lib/friends/introvert.rb:
```diff
@@ -777,5 +779,52 @@ def thing_with_name_in(type, text)
+        break if act.default_location != activity.default_location
+        consecutively_earliest_activity_with_same_default_location = act if act.default_location == activity.default_location
+      end
+      if consecutively_earliest_activity_with_same_default_location
```
In Ruby, the last line of a method is the method's return value, and whichever branch of a conditional gets evaluated is what the conditional evaluates to (`nil` if there's no `else` branch and none of the conditional branches are evaluated). So this could just be:

```ruby
if consecutively_earliest_activity_with_same_default_location
  consecutively_earliest_activity_with_same_default_location.date
else
  activity.date
end
```

or

```ruby
if consecutively_earliest_activity_with_same_default_location
  return consecutively_earliest_activity_with_same_default_location.date
end

activity.date
```

or

```ruby
(consecutively_earliest_activity_with_same_default_location || activity).date
```

shen-sat replied:
Ooo I like the last line, much shorter and more readable 😍 

I've seen people use `return` in other apps, I think it is to be super explicit with what is happening in the code. I guess if the code is a little busier, then there is a case for using it? Otherwise (like in this method), the return value is obvious 👍

JacobEvelyn commented on lib/friends/introvert.rb:
```diff
@@ -777,5 +779,52 @@ def thing_with_name_in(type, text)
+
+    def next_activity_date_with_different_default_location(activity, existing_activities)
+      next_activity_with_different_default_location = existing_activities.select { |a| a.default_location && (a.date > activity.date) && a.default_location != activity.default_location }.first
+      if next_activity_with_different_default_location
```
A failed conditional evaluates to `nil`, so this can just be:

```ruby
if next_activity_with_different_default_location
  next_activity_with_different_default_location.date
end
```

or

```ruby
next_activity_with_different_default_location&.date
```

shen-sat replied:
Gotcha - I don't use the `&.` enough. Making a mental note to use it more :)

JacobEvelyn commented on test/commands/graph_spec.rb:
```diff
@@ -273,7 +273,7 @@
       it "matches tag case-insensitively and scales the graph" do
         stdout_only <<-OUTPUT
-Nov 2015 |█████∙∙∙∙∙∙∙∙∙∙∙∙∙∙∙|
+Nov 2015 |∙∙∙∙∙∙∙∙∙∙∙∙∙∙∙∙∙∙∙∙|
```
I don't think I understand why this needed to change.

shen-sat replied:
This was something I forgot to explain. And it could be a case of me being too precious about the test descriptions. 

When I removed `Marie's Diner` in favor of `Martha's Vineyard`, the line:
`**Grace Hopper** and I went to _LOCATION_. George had to cancel at the last minute` no longer had anything to do with food, and so I removed the `@food` tag from the line. The tests then had to change to reflect this change.

JacobEvelyn replied:
Makes sense to me; thanks!

JacobEvelyn commented on test/commands/graph_spec.rb:
```diff
@@ -294,7 +294,7 @@
         it "matches tag case-insensitively and does not scale graph" do
           stdout_only <<-OUTPUT
-Nov 2015 |█
+Nov 2015 |∙|
```
Same question here.

shen-sat replied:
See comment above

JacobEvelyn commented on test/commands/add/activity_spec.rb:
```diff
@@ -51,5 +51,267 @@
+      let(:content) do
+  <<-FILE
+### Activities:
+- #{proceeding_activity}
```
It took me a second to see that this was different from `preceeding_activity`. How would you feel about changing the name to something like `following_activity`?

shen-sat replied:
Ha yes I got confused sometimes too! I was rushing a bit with variable names 😅 . `following_activity` is perefct 👍

JacobEvelyn commented on test/commands/add/activity_spec.rb:
```diff
@@ -51,5 +51,267 @@
+      end
+
+      describe "when proceeding activities have no default location" do
+        let(:proceeding_activity) { "2019-01-01: Visted a cafe"}
```
Typo: `Visted`

JacobEvelyn commented on test/commands/add/activity_spec.rb:
```diff
@@ -51,5 +51,267 @@
+        let(:proceeding_activity) { "2019-01-01: Visted a cafe"}
+
+        describe "when preceding activities have no default location" do
+          let(:preceding_activity) { "1999-01-01: Visted a library"}
```
Typo: `Visted`

JacobEvelyn commented on test/commands/add/activity_spec.rb:
```diff
@@ -51,5 +51,267 @@
+      end
+
+      describe 'when preceding activities have no default locations' do
+        let(:preceding_activity) { "1999-01-01: Visted a cafe"}
```
Typo: `Visted`

JacobEvelyn commented on test/commands/add/activity_spec.rb:
```diff
@@ -51,5 +51,267 @@
     end
   end
 
+  describe "adding default location" do
```
These tests are **_AMAZING_**. 😻

shen-sat replied:
I'm so glad you think so! 🙌 😄 

I'm not gonna lie, I went into a bit of a wormhole when thinking about all the different permutaions of adding default-location activities in the past 🤯 I must have re-written the tests at least three times haha

I **wish** it was magically possible to get bug reports on gems - it'd be so cool to get a bug report on default location functionality and then really tighten the tests to cover every edge case.

JacobEvelyn commented:
@shen-sat just left a review! Happy to discuss anything, as always!

shen-sat commented:
Apologies @JacobEvelyn, crazy busy week at work! Thanks for reviewing, I'll check out your comments on the weekend 🙂

shen-sat commented on README.md:
````diff
@@ -339,8 +340,8 @@ $ friends add activity last Monday
 ```bash
-$ friends add activity "2018-11-01: Grace and I went to \Marie's Diner. \George had to cancel at the last minute."
-Activity added: "2018-11-01: Grace Hopper and I went to Marie's Diner. George had to cancel at the last minute."
+$ friends add activity "2018-11-01: Grace and I went to \Martha's Vineyard. \George had to cancel at the last minute."
````
Martha is still being escaped as a name in this example, no? Albeit an old-school name :)

shen-sat commented on README.md:
````diff
@@ -420,6 +421,26 @@ $ friends add nickname "Grace Hopper" "Amazing Grace"
+When an activity includes the phrase to \_LOCATION\_ (e.g., Took a plane to \_Paris\_), all future activities that have no explicit location will be associated with that location:
+
+```bash
+$ friends add activity Took a plane to _Paris_
````
That makes perfect sense, I hadn't thought about the "new location" message! I'll make the change

shen-sat commented on lib/friends/event.rb:
```diff
@@ -138,8 +138,8 @@ def includes_location?(location)
-    def moved_to_location
-      @description[/(?<=[mM]oved to _)\w[^_]*(?=_)/]
+    def default_location
+      @description[/(?<=to _)\w[^_]*(?=_)/]
```
Memoizing is a good idea, I'll make the change.

Out of curiosity - it's impossible to memoize nil results right? I always thought, in memoization,`nil` is the trigger to perform an operation to get a value

shen-sat commented on lib/friends/introvert.rb:
```diff
@@ -110,6 +110,8 @@ def add_activity(serialization:)
 
         @output << "Activity added: \"#{activity}\""
+
+        decide_default_location_output(activity) if activity.default_location
```
Much easier to read 👍

shen-sat commented on lib/friends/introvert.rb:
```diff
@@ -777,5 +779,52 @@ def thing_with_name_in(type, text)
     end
+
+    def decide_default_location_output(activity)
+      @activities = stable_sort(@activities)
```
I think you mentioned before that we can use `stable_sort` in case the user somehow manually edits the activities in the `friends.md` file out of order. And even though this is an edge case, I think it's worth taking a precaution against, only because order is so important to the functionality in this PR.

And actually, I now think we should try and include `stable_sort` wherever we have functionality that depends on the order of activities, because it is that important. For example, we could also include it in the [set_implicit_locations!](https://github.com/shen-sat/friends/blob/a7794d507302226a950c407c99feaa153cb63e3b/lib/friends/introvert.rb#L663) method.

And following this thinking even further, maybe we should just add the `stable_sort` when we read the file? That way we don't have to repeat `stable_sort` in different places in the code.

shen-sat commented on lib/friends/introvert.rb:
```diff
@@ -777,5 +779,52 @@ def thing_with_name_in(type, text)
+
+    def decide_default_location_output(activity)
+      @activities = stable_sort(@activities)
+      existing_activities = @activities - [activity]
```
Makes sense! I didn't think about where we were adding the activity to `@activities`

shen-sat commented:
Busy week again this week, apologies @JacobEvelyn. I've addressed some comments, replied to others, and haven't responded/addressed the rest. If you are ok with waiting a few more days I will address/reply to the outstanding comments.

JacobEvelyn commented:
No problem, @shen-sat! Just let me know when I should take another look! (As a head's up, I'll be doing some travel over the next few weeks but should be able to check here sporadically.)

I'm really looking forward to getting this over the line!

shen-sat commented on lib/friends/introvert.rb:
```diff
@@ -777,5 +779,52 @@ def thing_with_name_in(type, text)
+
+      str = "Default location"
+      if activity != @activities.first
+        str += " from #{earliest_consecutive_default_location_activity_date(activity, existing_activities)} to"
```
I like your implementation - definitely way easier to read. I thought using TDD invariably would lead me to your implementation- a learning point for me!

I've swapped out my code for yours, and yup the tests all pass 🙌 I've removed one bit from your code: `< later_activities.last.date` - it was in the line for deciding if an added activity was the latest or not. `!later_activities.empty?` already does thsi check, and the tests still pass. Let me know if it was in for another reason!

JacobEvelyn replied:
Ah, good catch—I think you're right that that part of the condition was superfluous!

shen-sat commented on lib/friends/introvert.rb:
```diff
@@ -777,5 +779,52 @@ def thing_with_name_in(type, text)
+    end
+
+    def last_default_location_same_as_added_default_location?(activity, existing_activities)
+      last_default_location_activity = existing_activities.select { |a| a.default_location && (a.date < activity.date) }.first
```
Agreed - I'm removing this method in favor of your implementation, so no need to change

shen-sat commented on lib/friends/introvert.rb:
```diff
@@ -777,5 +779,52 @@ def thing_with_name_in(type, text)
+
+    def last_default_location_same_as_added_default_location?(activity, existing_activities)
+      last_default_location_activity = existing_activities.select { |a| a.default_location && (a.date < activity.date) }.first
+      if last_default_location_activity
```
Agreed - I'm removing this method in favor of your implementation, so no need to change

shen-sat commented:
Hey @JacobEvelyn - have addressed your comments and pushed latest changes. Code is now ready for your next review :) I've put a few questions/clarifications as well for you to look at.

Even though I'm going with your implementation and I'm deleting a lot of my code, I want to point out that I really appreciate you pointing out refactoring I can do in the soon-to-be-deleted code - super useful learning opportunties for me so please continue them 👍 

The elephants in the room now are my rubocop errors 😅 A lot of my test descriptions are super long, but I'm struggling with making them shorter whilst also making them descriptive for future devs. Do you have any suggestions?

shen-sat commented:
Hi @JacobEvelyn I've had a look at your comments and they all make sense 👍 Will implement changes hopefully this weekend.

I'm leaning towrads including `stable_sort` in one place, but I'll have a proper think about it while I implement the other changes.

Haha, yes we are so close! I can see the light at the end of this long (but interesting!) PR tunnel 😎

shen-sat commented:
Hey @JacobEvelyn hope you're well! I'm self isolating from the coronavirus atm - it sucks, but the only good thing about the situation is that I can focus on my code projects distraction free 👨🏾‍💻

I've made the changes we discussed. Of note:
- I've put `stable_sort` into the `read_file` method in `Introvert`, so now the `@activities` are sorted before doing anything
- I've put in tests for `set_implicit_location!` and `default_activity_output` that test against an unordered `###Activities` list
- There is one test failing - I think it's a test to essentially make sure we print an unordered `###Activities` list as an unordered list, though I'm not sure why we'd want this behaviour now that I've moved `stable_sort` to make sure everything is sorted correctly. Would it be ok to delete this test?

shen-sat commented on lib/friends/introvert.rb:
```diff
@@ -719,7 +725,7 @@ def parse_line!(line, line_num:, state:)
       begin
         instance_variable_get("@#{stage.id}") << stage.klass.deserialize(line)
-      rescue => ex # rubocop:disable Style/RescueStandardError
+      rescue StandardError => ex
```
I'm not sure why Rubocop is picking this up now?

shen-sat replied:
Sorry, a little more detail: Rubocop seems to have deleted the comment _and_ added `      rescue StandardError => ex` sll on its own. I'm a little confused as to why it's doing this now rather than in other PRs, but if you're happy then I'm happy :)

shen-sat commented on test/commands/list/activities_spec.rb:
```diff
@@ -55,15 +55,16 @@
+- 2000-01-04: Got lunch with **Grace Hopper** and **George Washington Carver**. @food
+- 2000-01-03: Celebrated my birthday in _Paris_ with **Marie Curie**. @partying @food
+- 2000-01-02: Went to _Atlantis_.
+- 2000-01-01: Talked to **George Washington Carver** on the phone for an hour.
```
I changed these dates to make their chronolgy clearer

JacobEvelyn commented on test/commands/add/activity_spec.rb:
```diff
@@ -51,5 +51,301 @@
+- 2018-01-01: Went to _Berlin_
+- 2019-01-01: Went to _Paris_
+
+
```
Remove this extra newline?

JacobEvelyn commented on test/commands/add/activity_spec.rb:
```diff
@@ -51,5 +51,301 @@
+FILE
+        end
+
+        it  "prints 'Default location from " \
```
Would you mind changing this to something more specific to the behavior we're trying to test? It should sort of "finish the sentence" started by the `describe` description. I'm thinking something along the lines of:

```
describe "when activities are out of order" do
  ...

  it "uses the sorted order for determining output" do
    ...
  end
end
```

since the actual format of what it prints isn't what this test is getting at.

Does that make sense?

shen-sat commented:
Hey @JacobEvelyn  - all changes made! 

Provided the build passes on Travis CI and I get the thumbs up from you, I'll squash the commits and push one more time - then I think the most epic PR I have ever been a part of is ready to be merged 😄 

Glad you are self isolating where you can 👍 It seems like the way to beat this thing is to stop the spread as soon as possible. It's difficult though, especially when it means not seeing friends and family as frequently. The flip side is that because many people are working from home, we all have more time than ever to contribute to open source #silverlining. 

And working on `friends` has become a staple part of my coding life now, so it's something familiar to cling to in these strange times. After a small break I might pickup a new new issue or create a new one - but I'll message you about that separately 👍

JacobEvelyn commented:
This looks good to me, @shen-sat! 🎉 

Thanks for all the hard work here! Once you squash (or I can squash on my end if you prefer) I'll merge and release a new version!

This ticket ended up being a lot bigger than I ever expected, so a huge huge thanks 🙏 for all your patience and work throughout this process. I'm also really glad you're doing everything you can to fight the pandemic. I am too (didn't mean to imply earlier that I wasn't fully isolating!) and I agree it's just a hard thing that requires a lot of effort from all of us. I'm happy to have you spending your extra time on open source though and `friends` in particular, and I'm really eager to see what we can get done together!

You've been awesome. Keep it up!

JacobEvelyn commented:
Actually, I’ve never used the “squash and merge” button on my end! Do you mind if I try it?

shen-sat commented:
Thanks @JacobEvelyn! And as usual, you've been too awesome to work with - I feel like we are developing a good working 'flow' 😊 And please feel free to git squash and merge 👍 

Just a few of things I'd like to mention as part of some post-PR feedback:
- Please keep up your open, friendly communication style. I mentioned it before, and I'll mention it again: having someone open to discussion around things makes me feel so much more invested in the project
- the refactorings you suggest as part of your PR reviews are always super easy to follow. I especially apreciated the pseudo-code you wrote for this PR, it really helped me understand what you were looking to implement
- I get carried away sometimes because during our chats we come up with lots of interesting ideas. And if I've had too many coffees I get all excited and try to implement them all in one go 🙃 But I'm realising this isn't the most efficient way to work, so on future PRs I want to continue our super communicative style where we come up with lots of improvements, but I'd like to capture those improvements in separate PRs. Does that sound ok?

And if you have any feedback for me, please let me know!

I'm taking a break from `friends` for a bit while I try and help out with the coronavirus situation. I actually found a decent site where devs can help out: https://helpwithcovid.com/ so I'm going to sign up to some projects listed there. But I will be back for sure 🙌

JacobEvelyn commented:
Hey @shen-sat! Thanks so much for the feedback! I really appreciate your openness with me.

> But I'm realising this isn't the most efficient way to work, so on future PRs I want to continue our super communicative style where we come up with lots of improvements, but I'd like to capture those improvements in separate PRs. Does that sound ok?

Yep, that makes sense to me! I think I definitely could have done a better job constraining ideas to keep the scope small. I'll try to be mindful of that in the future.

> And if you have any feedback for me, please let me know!

Honestly, you've been great to work with. I really have enjoyed bouncing ideas off of you and seeing the way you approach code. I worry that I'm sometimes too heavy-handed with how I want things done and really appreciate your flexibility in that regard!

> I'm taking a break from friends for a bit while I try and help out with the coronavirus situation. I actually found a decent site where devs can help out: https://helpwithcovid.com/ so I'm going to sign up to some projects listed there. But I will be back for sure 🙌

Awesome. I'm so glad you're safe and doing what you can to help others. I'm also doing what I can, and look forward to hearing from you on the other side of all this. Stay healthy! ❤️ 

Also, I just released your change in version `0.48`, which you can get with `friends update`. I'm personally really looking forward to this feature as it will let me clean up a _lot_ of cruft in my `friends.md` file. Let me know how you find it, and hopefully we'll get some feedback from others as well!
